### PR TITLE
protobuf : add version 4.25.0

### PR DESCRIPTION
**Pull request by sophieeihpos**
protobuf/4.25.0

Part 1 fix for https://github.com/conan-io/conan-center-index/issues/23084

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

© 2024 Morgan Stanley.
THIS SOFTWARE IS CONTRIBUTED SUBJECT TO THE TERMS OF THE Conan-io project Contributor License Agreement.
THIS SOFTWARE IS LICENSED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE AND ANY WARRANTY OF NON-INFRINGEMENT, ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. THIS SOFTWARE MAY BE REDISTRIBUTED TO OTHERS ONLY BY EFFECTIVELY USING THIS OR ANOTHER EQUIVALENT DISCLAIMER IN ADDITION TO ANY OTHER REQUIRED LICENSE TERMS

### CODE REVIEW

**Comment by Ahajha** [comment]
Note: There's another PR to add 4.25.3 here: #21622

Also, note that protobuf's versioning is a little strange, there is no 3.22, it goes from 3.21 to 4.22. See here: https://protobuf.dev/support/version-support/#cpp

File: recipes/protobuf/all/conanfile.py
Comment (Ahajha) [inline recipes/protobuf/all/conanfile.py:266]: Also here, what's in the utf8_range folder? Why is it being removed?

File: recipes/protobuf/all/conanfile.py
Comment (Ahajha) [inline recipes/protobuf/all/conanfile.py:242]: Last thing, I'm curious about all of the RPATH workarounds, why are they necessary? Do these changes improve the situation? I would hope that a VirtualRunEnv/VirtualBuildEnv would allow them to work unmodified, is that not the case?

**Review by Ahajha — CHANGES_REQUESTED** [review]

File: recipes/protobuf/all/conanfile.py
Comment (sophieeihpos) [reply recipes/protobuf/all/conanfile.py:85]: The cppstd is checked at compile time, so I don't have a better solution if it is not defined by default.

File: recipes/protobuf/all/conanfile.py
Comment (sophieeihpos) [reply recipes/protobuf/all/conanfile.py:277]: protobuf-generate is a file renamed renamed from protobuf-config in the current CCI. The later versions already have a protobuf-generate by default.

File: recipes/protobuf/all/conanfile.py
Comment (sophieeihpos) [reply recipes/protobuf/all/conanfile.py:266]: cmake config files for utf8_range. They were checked by the hook.

File: recipes/protobuf/all/conanfile.py
Comment (Ahajha) [reply recipes/protobuf/all/conanfile.py]: Would it not be defined by the CMake build system to be 11 or 14 if it needs to? I don't expect not defining it to break anything. A ton of other CCI recipes use this pattern, but there's no `else` there. Maybe I'm misunderstanding the problem though?

Comment (sophieeihpos) [reply recipes/protobuf/all/conanfile.py]: check_min_cppstd on its own without the "if check" seems to be working fine.

File: recipes/protobuf/all/conanfile.py
Comment (Ahajha) [reply recipes/protobuf/all/conanfile.py:238]: I agree that first part is a pain point, but my main concern with this change now that I've looked into how it works is that I don't know if the packages will be relocatable. Mainly since you use the relative path between two package folders. Maybe @uilianries has some insight?

File: recipes/protobuf/all/conanfile.py
Comment (Ahajha) [reply recipes/protobuf/all/conanfile.py]: Still looks like all of the v1 checks are invalid. There's a line in the implementation that checks for cppstd to be defined:
```python
    current_cppstd = conanfile.settings.get_safe("compiler.cppstd")
    if current_cppstd is None:
        raise ConanInvalidConfiguration("The compiler.cppstd is not defined for this configuration")
```
So I think that check still needs to be guarded, and the v1 builds fixed.

Comment (sophieeihpos) [reply recipes/protobuf/all/conanfile.py]: The doc for check_min_cppstd says: "Validates if the applied cppstd setting (from compiler.cppstd settings or deducing the default from compiler and compiler.version) is at least the value specified in the cppstd argument. "

The v1 builds have been run already ?

File: recipes/protobuf/all/conanfile.py
Comment (sophieeihpos) [reply recipes/protobuf/all/conanfile.py:238]: The LD_LIBRARY_PATH will be searched for, if dependency binaries are not found in its rpath.

File: recipes/protobuf/all/conanfile.py
Comment (Ahajha) [reply recipes/protobuf/all/conanfile.py]: The check only happens if the cppstd is defined, and the implementation suggests that if it isn't, the package is automatically invalid.

The v1 checks pass, but because nothing ends up being built.

File: recipes/protobuf/all/conanfile.py
Comment (Ahajha) [reply recipes/protobuf/all/conanfile.py:238]: Personally I would prefer an option, but since I understand now that it would work either way, I'm not sure. I think I'll defer to a maintainer, but as-is I think it's okay for now.

File: recipes/protobuf/all/conanfile.py
Comment (sophieeihpos) [reply recipes/protobuf/all/conanfile.py]: The v1 build was not skipped : https://c3i.jfrog.io/c3i/misc/logs/pr/23095/39-linux-gcc-legacy/protobuf/3.21.12//7627fae1426bcc12a67dba7c7207b1bccf05e5fd-build.txt

Comment (Ahajha) [reply recipes/protobuf/all/conanfile.py]: Specifically the checks for the new version were skipped, the others are run because of the version check above: https://c3i.jfrog.io/c3i/misc/summary.html?json=https://c3i.jfrog.io/c3i/misc/logs/pr/23095/39/protobuf/4.25.0//summary.json

Comment (Ahajha) [reply recipes/protobuf/all/conanfile.py]: Sorry about any confusion!

I think the build systems for each should correctly set the C++ standard, it's just that Conan (in the v1 pipeline) doesn't know what it is, so it can't check if it's valid. I believe that's the reason it's recommended in Conan v2. So I think just have the check be guarded by `if self.settings.compiler.get_safe("cppstd"):` should be sufficient.

Comment (sophieeihpos) [reply recipes/protobuf/all/conanfile.py]: One of the compiler failed for the check `#if __cplusplus < 201402L` at build time.

**Comment by HennerM** [comment]
@sophieeihpos thanks for working on this! Do you know when this can get merged?

**Comment by Ahajha** [comment]
Just as an FYI for reviewers, there is another PR open here: https://github.com/conan-io/conan-center-index/pull/21622.

**Comment by sophieeihpos** [comment]
Any estimate when this PR or the other will get reviewed ? @Ahajha @uilianries

**Comment by Ahajha** [comment]
I'll try to re-review in the next day or two, sorry for the delay. (Will still need two other reviews). I did glance at it a few days ago and everything seemed okay, my only immediate concern is that a maintainer might request that the new option you've added go in a separate PR, but I'm not sure on the policy myself. Adding some more context to the changes in the PR description also tends to help things along I find.

File: recipes/protobuf/all/test_package/conanfile.py
Comment (Ahajha) [inline recipes/protobuf/all/test_package/conanfile.py:24]: I realized there's a better way of doing this, since in the test package we're in control of the CMake version, in the CMake we can add the following:
```cmake
if (protobuf_VERSION VERSION_LESS "4.0.0")
    target_compile_features(${PROJECT_NAME} PRIVATE cxx_std_11)
else()
    target_compile_features(${PROJECT_NAME} PRIVATE cxx_std_14)
endif()
```
In place of the existing `target_compile_features`.

**Review by Ahajha — APPROVED** [review]
Overall LGTM - other reviewers keep in mind there is another PR so maybe there's some stuff in that one that might be worth looking at.

**Comment by sophieeihpos** [comment]
> I'll try to re-review in the next day or two, sorry for the delay. (Will still need two other reviews). I did glance at it a few days ago and everything seemed okay, my only immediate concern is that a maintainer might request that the new option you've added go in a separate PR, but I'm not sure on the policy myself. Adding some more context to the changes in the PR description also tends to help things along I find.

Thank you for the review. Any ideas where I can get more reviewers ?

File: recipes/protobuf/all/conanfile.py
Comment (sophieeihpos) [reply recipes/protobuf/all/conanfile.py:78]: For 4.25.0+, shared windows build is not supported. Shared windows abseil is not available in its recipe, and static windows abseil does not build with shared protobuf.

File: recipes/protobuf/all/test_package/conanfile.py
Comment (sophieeihpos) [reply recipes/protobuf/all/test_package/conanfile.py:24]: I will try and see.

**Comment by sophieeihpos** [comment]
@RubenRBS @uilianries Could you please help review ?

**Comment by saschasc** [comment]
Is there any advantage adding 4.25.0 if there exists a PR for 4.25.3?

https://github.com/conan-io/conan-center-index/pull/21622

Anyway would be great if some of the reviewers could push this. Thank you.

**Comment by jcar87** [comment]
Thanks @sophieeihpos  -  closing as superseded by https://github.com/conan-io/conan-center-index/pull/24154


It's unclear from this PR what problems the proposed rpath logic are trying to fix - if this is still a concern, could you please open a new issue providing more details? if this is still a concern, could you please open a new issue providing more details? We will be happy to assist, thank you so much!

**Pull request closed without merging**
